packages/react-dom/src/__tests__/ReactTreeTraversal-test.js

```diff
@@ -0,0 +1,304 @@
+    }
+    onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+    onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}>
+    <div
+      id={id + '__DIV_1'}
+      onClickCapture={e =>
+        eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+      }
+      onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+      onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    />
+    <div
+      id={id + '__DIV_2'}
+      onClickCapture={e =>
+        eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+      }
+      onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+      onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG2)}
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    />
+  </div>
+);
+
+const ParentComponent = ({eventHandler}) => (
+  <div
+    id="P"
+    onClickCapture={e =>
+      eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+    }
+    onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+    onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}>
+    <div
+      id="P_P1"
+      onClickCapture={e =>
+        eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+      }
+      onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+      onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}>
+      <ChildComponent id="P_P1_C1" eventHandler={eventHandler} />
+      <ChildComponent id="P_P1_C2" eventHandler={eventHandler} />
+    </div>
+    <div
+      id="P_OneOff"
+      onClickCapture={e =>
+        eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
+      }
+      onClick={e => eventHandler(e.currentTarget.id, 'bubbled', e.type, ARG)}
+      onMouseEnter={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+      onMouseLeave={e => eventHandler(e.currentTarget.id, e.type, ARG)}
+    />
+  </div>
+);
+
+describe('ReactTreeTraversal', () => {
+  var mockFn = jest.fn();
+  var container;
+
+  beforeEach(() => {
+    React = require('react');
+    ReactDOM = require('react-dom');
+
+    mockFn.mockReset();
+
+    container = document.createElement('div');
+    document.body.appendChild(container);
+
+    ReactDOM.render(<ParentComponent eventHandler={mockFn} />, container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+    container = null;
+  });
+
+  describe('Two phase traversal', () => {
+    it('should not traverse when target is outside component boundary', () => {
+      const outerNode = document.createElement('div');
+      document.body.appendChild(outerNode);
```

Comment:
The node is appended here to the document. We should remove it to clean up after the test. I usually put the rest of the test in try/finally, and remove it in the finally block.